### ISSUE TITLE
Get and use dynamic 24h%change from all informative pairs in the strategy.

### ISSUE THREAD

**Issue by n8an** [freqtrade/freqtrade#7048]
## Describe your environment

  * Operating system: Docker Image
  * Freqtrade Version: 2022.6
  
## Your question

What is the best way to get all informative pairs last 24hour % change, have it be dynamic (just like on the exchange itself, where it updates the last 24h%change every minute), and use this dynamic data from all informative pairs in the strategy as a signal for example.

**Comment by n8an** [comment]
All implementations I have in mind would be too slow, since they require to load all 24 hour 1minute data for all pairs every minute to calculate their exact 24h% change.
I'm also trying to use ta.ROC now for this, but still in process of getting it to work.

**Comment by xmatthias** [comment]
I doubt using 1m candles makes sense for such a thing ... (you'll get ~1000 candles by default for most exchanges - so you'll never get 24h anyway) unless an exchange provides more data in one call (ftx for example).

i don't think ROC is correct - you'll really want to do "last row" vs. "last row - 1440 rows" in your comparison ...

performance will suffer from this - but that'll always be the case when looking at 1m data - where i always advise to reconsider - "do you really need 1m data - or would higher level data suffice".

Think about it this way - a jump in 24h price can have 2 reasons - either the price is jumping NOW - or it was jumping 24h ago (it's not an average that's changing ...).
so you either act on "now" data - or on data 1 day old. You'll however not know (without looking at other, already available data) where your signal came from (today or yesterday).

**Comment by n8an** [comment]
Well, I'm going to use this 24hour percent change not as a signal, but as a part of a signal.
I'm just looking for a best way to get the precise minute to minute (almost live) data on all cryptocurrencies from whitelist with their 24h%change.
Right now trying to use the dataprovider and ta.roc to get the desired result.
I guess the only way for this to be fast is to simply get the "24h%change" from the exchange itself when the data is dowloaded.
But I don't recall if Binance provides this.

Basically I need to check all pairs 24h percentages every minute.
If there is an elegant way to do this with freqtrade, built in functions, plugins may be, I would love to know.

**Comment by xmatthias** [comment]
it MAY be part of the [ticker output](https://www.freqtrade.io/en/stable/strategy-customization/#tickerpair) (24h volume is) - but i'm not sure if binance provides the "24h change" as well or not.

What they provide may also vary between spot and futures markets - so such a strategy will not easily transfer between markets or exchanges.

When using this function though - it is not backtestable - so have some "fallbacks" for that in mind if you need backtesting support.

**Comment by n8an** [comment]
Yes, I have thought of using simple ticker output from binance if it provides the 24h change. But of course only for live strategy.
But since right now I need to backtest everything, seems like the only way to get what I want, is by doing it the way I described below.

For backtesting
1. download all whitelisted pairs 1m data for the duration of backtest
2. in the strategy, use all whitelisted pairs as informative
3. apply to all of them the TA.ROC with desired 24h length to get % change
4. use this data to create other indicators, or entry or exit signals
5. backtest

(one of the thing I'm unsure about is, in where the best to do the 3rd step. inside of the "informative_pairs" function or "populate_indicators")

If I'm correct in this assumption, please confirm, or disprove it.
And of course feel free to correct me if I'm totally wrong.

**Comment by n8an** [comment]
Also for some reason the "merge_informative_pair()" function doesn't exist for me.

```
freqtrade 2022.6

2022-07-07 13:30:39,581 - freqtrade.exchange.exchange - INFO - Using CCXT 1.89.14

Traceback (most recent call last):
  File "/freqtrade/freqtrade/main.py", line 37, in main
    return_code = args['func'](args)
  File "/freqtrade/freqtrade/commands/optimize_commands.py", line 58, in start_backtesting
    backtesting.start()
  File "/freqtrade/freqtrade/optimize/backtesting.py", line 1256, in start
    min_date, max_date = self.backtest_one_strategy(strat, data, timerange)
  File "/freqtrade/freqtrade/optimize/backtesting.py", line 1154, in backtest_one_strategy
    preprocessed = self.strategy.advise_all_indicators(data)
  File "/freqtrade/freqtrade/strategy/interface.py", line 1127, in advise_all_indicators
    return {pair: self.advise_indicators(pair_data.copy(), {'pair': pair}).copy()
  File "/freqtrade/freqtrade/strategy/interface.py", line 1127, in <dictcomp>
    return {pair: self.advise_indicators(pair_data.copy(), {'pair': pair}).copy()
  File "/freqtrade/freqtrade/strategy/interface.py", line 1145, in advise_indicators
    return self.populate_indicators(dataframe, metadata)
  File "/freqtrade/user_data/strategies/MyStrategy.py", line 178, in populate_indicators
    dataframe = merge_informative_pair(dataframe, informative, self.timeframe, inf_tf, ffill=True)
NameError: name 'merge_informative_pair' is not defined
ERROR: 1
```

**Comment by xmatthias** [comment]
You're probably missing an import at the top of the file. The function will not exist magically.

you may want to do the merge (assuming i understood it right - for ALL pairs combined - for each pair) - `bot_loop_start`. otherwise the whole combination will run for every pair - which is kindof a waste as it won't change.
You can "cache" it in a variable (self.whatever) - and then have access to it in subsequent methods.

if you'd just use one  (or a few) informative pairs for each pair - then I'd recommend to go the [`@informative` decorator](https://www.freqtrade.io/en/stable/strategy-customization/#informative-pairs-decorator-informative) way. it's the simplest way and will cover 2 and 3

**Comment by n8an** [comment]
It seems that I have found a way to do it, but now I have a problem with the whitelist itself.

```
docker-compose run --rm freqtrade backtesting -v --strategy MyStrategy -p btc/usdt
```

When using the command above for backtesting with the pair I want, the whitelist becomes this pair alone.
Which is not desired.
What is the right way to have backtesting done on 'btc/usdt' alone, but having the whole whitelist for informative pairs?
Right now the only idea I've got is just manually loading in the 'whitelist' I want into informative pairs function.

**Comment by xmatthias** [comment]
you can't (not if you use whitelist for informative as it's in the docs). 
whitelist is what defines what is backtested - so `-p` really justs writes what's there to the whitelist.
That's expected and desired.

What you want can only be accomplished by either backtesting all pairs - or hard-coding the desired informative pair list (it's now no longer a whitelist) into the informative section of the strategy.

**Comment by n8an** [comment]
I figured the same.
Does the merge_imformative_pairs() override the data if it's called multiple times?

```
                           date     open     high      low    close   volume       roc                   date_1m  open_1m  high_1m  low_1m  close_1m  volume_1m  ZRX/USDT_roc_1m  enter_long  enter_short
43193 2022-07-06 15:43:00+00:00  20085.0  20122.9  20085.0  20091.8  405.302  0.033358 2022-07-06 15:43:00+00:00   0.3026   0.3030  0.3026    0.3027     9877.0         2.022245         NaN          NaN
43194 2022-07-06 15:44:00+00:00  20091.9  20109.2  20087.6  20087.7  149.368 -0.020406 2022-07-06 15:44:00+00:00   0.3027   0.3030  0.3026    0.3027    31363.4         1.953520         NaN          NaN
43195 2022-07-06 15:45:00+00:00  20087.6  20107.5  20080.0  20107.4  279.537  0.098070 2022-07-06 15:45:00+00:00   0.3026   0.3030  0.3026    0.3030    30526.6         1.985863         NaN          NaN
43196 2022-07-06 15:46:00+00:00  20107.4  20123.3  20100.0  20119.9  352.837  0.062166 2022-07-06 15:46:00+00:00   0.3030   0.3034  0.3030    0.3032    26181.2         2.087542         NaN          NaN
43197 2022-07-06 15:47:00+00:00  20119.8  20119.8  20103.3  20108.7  150.943 -0.055666 2022-07-06 15:47:00+00:00   0.3032   0.3033  0.3028    0.3031    18037.0         2.364066         NaN          NaN
```

You can see that ZRX/USDT is the data column from the merge, and it's the last one on the list of pairs I have, so looks like the merge function doesn't add the data to the right, it overrides it every time.

**Comment by xmatthias** [comment]
Depends how you implement it - the method itself really only does a slightly more intelligent `pd.merge()`.
columns are renamed to be postfixed with the timeframe - but that's it. If duplicate columns appear (you have the expected outcome column already in the input dataframe) - then sure, it'll overwrite that.


You can work around this pretty easily by prefixing the "required" columns with a name (like you did with ZRX) This gives you full control over which columns you require - and how they'll be named.

It all depends how it's used, really - the method doesn't do this automatically for you - but i don't think it should either, as this can be expected (you may not want all ohlc columns - but only "what you're interested in").

**Comment by n8an** [comment]
This is how it's done in my code, and it still overrides everything for some reason.
```
def populate_indicators(self, dataframe: DataFrame, metadata: dict) -> DataFrame:
        if not self.dp:
            return dataframe

        dataframe['roc'] = ta.ROC(dataframe, timeperiod=1)

        for pairlist in self.informative_pairs():
            inf_pair, inf_tf = pairlist
            inf_data = self.dp.get_pair_dataframe(pair=inf_pair, timeframe=inf_tf)
            inf_data[f'{inf_pair}_roc'] = ta.ROC(inf_data, timeperiod=1440)
        
        dataframe = merge_informative_pair(dataframe, inf_data, self.timeframe, inf_tf, ffill=True)

        print(inf_data.tail())

        return dataframe
```
_dataframe = merge_informative_pair(dataframe, **inf_data**, self.timeframe, inf_tf, ffill=True)_
is outside of the for loop here

And also I have tried this.
```
def populate_indicators(self, dataframe: DataFrame, metadata: dict) -> DataFrame:
        if not self.dp:
            return dataframe

        dataframe['roc'] = ta.ROC(dataframe, timeperiod=1)

        for pairlist in self.informative_pairs():
            inf_pair, inf_tf = pairlist
            inf_data = self.dp.get_pair_dataframe(pair=inf_pair, timeframe=inf_tf)
            inf_data[f'{inf_pair}_roc'] = ta.ROC(inf_data, timeperiod=1440)
            dataframe = merge_informative_pair(dataframe, inf_data[f'{inf_pair}_roc'], self.timeframe, inf_tf, ffill=True)

        print(inf_data.tail())

        return dataframe
```
_dataframe = merge_informative_pair(dataframe, **inf_data[f'{inf_pair}_roc']**, self.timeframe, inf_tf, ffill=True)_
is inside and uses **inf_data[f'{inf_pair}_roc']**

Both still result in override, not too sure how.

**Comment by xmatthias** [comment]
you're getting a new dataframe in every loop - so the first is clearly, completely wrong.

The 2nd has more potential - but i don't undrestand why you're passing in the column only - it requires the dataframe, so most likely, it'll still not work.

**Comment by n8an** [comment]
Is there a way, to merge_informative, but without adding dates as new columns, using the original dates only?

**Comment by xmatthias** [comment]
it's designed to work for most people "out of the box". this means there's also not much configurability.

The [docs](https://www.freqtrade.io/en/stable/strategy-customization/#merge_informative_pair) about it do however contain a segment about how to implement this yourself - where you can decide yourself how to process things, what to keep, ...
